### PR TITLE
[coop] Add handles for some SRE scenarios

### PR DESCRIPTION
**Pull request by BrzVlad**

### CODE REVIEW

File: src/mono/mono/metadata/handle.h [hunk 1]
```diff
@@ -517,7 +517,10 @@ TYPED_HANDLE_DECL (MonoException);
 TYPED_HANDLE_DECL (MonoAppContext);
 
 /* Simpler version of MONO_HANDLE_NEW if the handle is not used */
-#define MONO_HANDLE_PIN(object) MONO_HANDLE_NEW (MonoObject, (MonoObject*)(object))
+#define MONO_HANDLE_PIN(object) do { \
+		if ((object) != NULL) \
```
Comment (jaykrell) [inline src/mono/mono/metadata/handle.h:521]: I don't think the null check is needed.

File: src/mono/mono/metadata/sre-save.c [hunk 2]
```diff
@@ -685,26 +689,34 @@ mono_image_get_method_info (MonoReflectionMethodBuilder *mb, MonoDynamicImage *a
...
+	HANDLE_FUNCTION_RETURN_VAL (ret);
 }
 
 static gboolean
 mono_image_get_ctor_info (MonoDomain *domain, MonoReflectionCtorBuilder *mb, MonoDynamicImage *assembly, MonoError *error)
 {
+	HANDLE_FUNCTION_ENTER ();
 	MONO_REQ_GC_UNSAFE_MODE;
 
+	gboolean ret;
```
Comment (jaykrell) [inline src/mono/mono/metadata/sre-save.c:703]: Maybe initialize it to `FALSE`?

File: src/mono/mono/metadata/sre-save.c [hunk 3]
```diff
@@ -630,17 +630,21 @@ static gboolean
 mono_image_get_method_info (MonoReflectionMethodBuilder *mb, MonoDynamicImage *assembly, MonoError *error)
 {
 	MONO_REQ_GC_UNSAFE_MODE;
+	HANDLE_FUNCTION_ENTER ();
 
 	MonoDynamicTable *table;
 	guint32 *values;
 	ReflectionMethodBuilder rmb;
 	int i;
+	gboolean ret;
```
Comment (jaykrell) [inline src/mono/mono/metadata/sre-save.c:639]: Initialize to `FALSE`?

File: src/mono/mono/metadata/sre.c [hunk 4]
```diff
@@ -3262,38 +3290,47 @@ reflection_methodbuilder_to_mono_method (MonoClass *klass,
 static MonoMethod*
 ctorbuilder_to_mono_method (MonoClass *klass, MonoReflectionCtorBuilder* mb, MonoError *error)
 {
+	HANDLE_FUNCTION_ENTER ();
 	ReflectionMethodBuilder rmb;
 	MonoMethodSignature *sig;
+	MonoMethod *ret;
```
Comment (jaykrell) [inline src/mono/mono/metadata/sre.c:3296]: Initialize to `NULL`?

File: src/mono/mono/metadata/sre-save.c [hunk 5]
```diff
@@ -630,17 +630,21 @@ static gboolean
 mono_image_get_method_info (MonoReflectionMethodBuilder *mb, MonoDynamicImage *assembly, MonoError *error)
 {
 	MONO_REQ_GC_UNSAFE_MODE;
+	HANDLE_FUNCTION_ENTER ();
```
Comment (vargaz) [inline src/mono/mono/metadata/sre-save.c:634]: Why is this needed if this function doesn't create any handles ?

Comment (BrzVlad) [reply src/mono/mono/metadata/sre-save.c:634]: This method owns the `ReflectionMethodBuilder` and all the pinned handles created from the initialization of the structure. `mono_reflection_methodbuilder_from_method_builder` allocates the handles which live past the method  return.

File: src/mono/mono/metadata/handle.h (hunk 1 above)
Comment (BrzVlad) [reply src/mono/mono/metadata/handle.h:521]: Some object fields of the `ReflectionMethodBuilder` can be null. Allocating a handle in that case is pointless. This avoid doing explicit checks for each field.

File: src/mono/mono/metadata/sre-save.c (hunk 5 above)
Comment (vargaz) [reply src/mono/mono/metadata/sre-save.c:634]: Could you add some comments for that ? This is not the usual way handles are used.

**Review by vargaz — APPROVED** [review]
